### DISCUSSION TITLE
Single threaded in production

### DISCUSSION THREAD

**Discussion by singpolyma** [socketry/falcon#165] [Q&A] [answered]
We are currently used to using `thin` via `rackup` in production for single-threaded but evented (and thus concurrent) request handling with eventmachine.

We are curious about leaving eventmachine for ruby3. The documentation says that using falcon with `rackup` is not recommended, but is this just because it will be single-threaded?  Since our code will all be async, that seems like not such a big deal and probably we can expect performance similar to thin when used in this way?

**Comment by ioquatix** [comment]
How many processor cores do you have on the instance and do you want to use them all?

**Comment by singpolyma** [comment]
I'm aware that a single ruby process won't use multiple cores, but our webapps are generally I/O bound not CPU bound so that's never been a concern.

**Comment by ioquatix** [comment, marked as answer]
The `rackup` command has a number of limitations.

- Limited control over server multi-process/multi-thread configuration.
- Limited control over protocol selection.
- Limited control over binding address (and frankly pretty strange options for handling it).
- Limited fault tolerate (e.g. app crash / restart process).
- Limited support for rolling restart.

If you've tested your use case and none of these issues bother you, it might be an acceptable solution to use `rackup`.
